static/skin/index.js

```diff
@@ -207,34 +207,42 @@
         const zimUrl = params.get('as'); // as= is fallback URL
         // download metalink to build list of mirrored URLs
         let mirrorUrls = [];
-        try {
-            const metalink = await fetch(`${zimUrl}.meta4`).then(response => {
-                return response.ok ? response.text() : '';
-            }).catch((_error) => '');
-            if (metalink) {
+
+        const metalink = await fetch(`${zimUrl}.meta4`).then(response => {
+            return response.ok ? response.text() : '';
+        }).catch((_error) => '');
+        if (metalink) {
+            try {
                 const parser = new DOMParser();
                 const doc = parser.parseFromString(metalink, "application/xml");
                 doc.querySelectorAll("url").forEach((node) => {
                     if (node.hasAttribute("priority")) { // ensures its a mirror link
                         mirrorUrls.push(node.innerHTML);
                     }
                 });
+            } catch (err) {
+                // not a big deal, magnet will only contain primary URL
+                console.debug(`Failed to parse mirror links for ${zimUrl}`);
             }
-        } catch (err) {
-            // not a big deal, magnet will only contain primary URL
-            console.debug(`Failed to retrieve mirror links for ${zimUrl}`);
         }
 
         // set webseed (ws=) URL to primary download URL (redirects to mirror)
-        params.set('ws', zimUrl);
+        params.set('ws', encodeURIComponent(zimUrl));
         // if we got metalink mirror URLs, append them all
         if (mirrorUrls) {
             mirrorUrls.forEach((url) => {
-                params.append('ws', url);
+                params.append('ws', encodeURIComponent(url));
             });
         }
 
-        params.set('xs', `${zimUrl}.torrent`);  // adding xs= to point to torrent URL
+        params.set('as', encodeURIComponent(zimUrl));
+        params.set('dn', encodeURIComponent(params.get('dn')));
+        params.set('xs', encodeURIComponent(`${zimUrl}.torrent`));  // adding xs= to point to torrent URL
+        const trackers = params.getAll('tr');
+        params.delete('tr');
+        trackers.forEach((tracker) => {
+            params.append('tr', encodeURIComponent(tracker));
+        });
```

Comment:
Isn't it better to let `URLSearchParams.toString()` do the URI-encoding?

Reply:
No, see comment above

Reply:
Then why don't you instead URI-encode inside the loop where you build the final link?

Reply:
Because only URLs containing parts needs `encodeURIComponent` while others would not support it. We'd have to toggle with it and nothing or `encodeURI`. In current implementation, it's very clear what requires to be encoded IMO

Reply:
IMHO, it will be clearer if there is a list (array) of keys requiring URI-encoding and a single call to `encodeURIComponent()` in a loop.

Reply:
I would take that approach by introducing a reusable function `makeURLSearchString(params, keysToURIEncode)`